File: GenesisAndroid/src/main/java/com/emerchantpay/gateway/genesisandroid/api/internal/response/Response.java
package com.emerchantpay.gateway.genesisandroid.api.internal.response;

import com.emerchantpay.gateway.genesisandroid.api.constants.TransactionStates;
import com.emerchantpay.gateway.genesisandroid.api.internal.GenesisClient;
import com.emerchantpay.gateway.genesisandroid.api.models.GenesisError;
import com.emerchantpay.gateway.genesisandroid.api.models.Transaction;
import com.emerchantpay.gateway.genesisandroid.api.util.NodeWrapper;

public class Response {

    // Genesis client
    private GenesisClient client;

    public Response(GenesisClient client) {
        super();

        this.client = client;
    }

    // Get Redirect Url
    public String getRedirectUrl() {
        return getTransaction().getRedirectUrl();
    }

    // Get Unique Id
    public String getUniqueId() {
        return getTransaction().getUniqueId();
    }

    // Get status
    public String getStatus() {
        return getTransaction().getStatus();
    }

    // Get Status code
    public Integer getCode() {
        return getTransaction().getCode();
    }

    // Get Technical message
    public String getTechnicalMessage() {
        String technicalMessage = getTransaction().getTechnicalMessage();

        if (technicalMessage == null || technicalMessage.isEmpty()) {
            NodeWrapper paymentTransaction = getTransaction().getPaymentTransactions().get(0);
            return paymentTransaction.findString("technical_message");
        } else {
            return technicalMessage;
        }
    }

    // Get Message
    public String getMessage() {
        return getTransaction().getMessage();
    }

    public Transaction getTransaction() {
        return client.getTransaction().getRequest().getTransaction();
    }

    // Get success transaction
    public Boolean isSuccess() {
        if (getStatus().equals(TransactionStates.ERROR) ||
                getStatus().equals(TransactionStates.DECLINED)) {
            return false;
        } else {
            return true;
        }
    }

    // Get Error object
    public GenesisError getError() {
        if (!isSuccess()) {
            return new GenesisError(getCode(), getMessage(), getTechnicalMessage());
        } else {
            return null;
        }
    }
}

File: GenesisAndroid/src/main/java/com/emerchantpay/gateway/genesisandroid/api/internal/response/Response.kt
package com.emerchantpay.gateway.genesisandroid.api.internal.response

import com.emerchantpay.gateway.genesisandroid.api.constants.TransactionStates
import com.emerchantpay.gateway.genesisandroid.api.internal.GenesisClient
import com.emerchantpay.gateway.genesisandroid.api.models.GenesisError
import com.emerchantpay.gateway.genesisandroid.api.models.Transaction

open class Response(// Genesis client
        private val client: GenesisClient) {

    // Get Redirect Url
    val redirectUrl: String?
        get() = transaction?.redirectUrl

    // Get Unique Id
    val uniqueId: String?
        get() = transaction?.uniqueId

    // Get status
    val status: String?
        get() = transaction?.status

    // Get Status code
    val code: Int?
        get() = transaction?.code

    // Get Technical message
    val technicalMessage: String
        get() {
            val technicalMessage = transaction?.technicalMessage

            return when {
                technicalMessage == null || technicalMessage.isEmpty() -> {
                    val paymentTransaction = transaction?.paymentTransactions?.get(0)
                    paymentTransaction?.findString("technical_message")
                }
                else -> technicalMessage
            } as String
        }

    // Get Message
    val message: String?
        get() = transaction?.message

    val transaction: Transaction?
        get() = client.transaction?.request?.transaction

    // Get success transaction
    val isSuccess: Boolean?
        get() = !(status == TransactionStates.ERROR || status == TransactionStates.DECLINED)

    // Get Error object
    val error: GenesisError?
        get() = when {
            (!isSuccess!!)!! -> message?.let { GenesisError(code, it, technicalMessage) }
            else -> null
        }
}